File: java/ankit_kanyal/devnest_problems/phase_2/TreeBoundary.java
// link- https://devsnest.in/algo-challenges/tree-boundary?tab=question

static boolean isLeaf(TreeNode root){
    return root.left==null && root.right==null;
}

static void addLeaves(TreeNode root, List<Integer> result) {
        if (isLeaf(root)) {
            result.add(root.val);
            return;
        }
        if (root.left!=null) addLeaves(root.left, result);
        if (root.right!=null) addLeaves(root.right, result);
    }

static int[] solve(TreeNode root){
//CODE HERE 
if(root==null) return new int[]{};
List<Integer> result=new ArrayList<>();

if(!isLeaf(root)) result.add(root.val);
// for left boundary

TreeNode cur=root.left;

while(cur!=null){
    if(!isLeaf(cur)) result.add(cur.val);
    if(cur.left!=null) cur=cur.left;
    else cur=cur.right;
}

// add leaves nodes
cur=root.right;
addLeaves(root,result);

Stack<Integer> stack=new Stack<>();
while(cur!=null){
    if(!isLeaf(cur)) stack.push(cur.val);
    if(cur.right!=null) cur=cur.right;
    else cur=cur.right;
}

while(!stack.empty()){
    result.add(stack.pop());
}

return result.stream().mapToInt(Integer::intValue).toArray();

}

File: Python/chandu/Tree_Boundary.py
def solve(root):
    # CODE HERE
    if root is None:
        return []
    kat = [root.val]

    def leftNode(node):
        if not node or not node.left and not node.right:
            return
        kat.append(node.val)
        if node.left:
            leftNode(node.left)
        else:
            leftNode(node.right)
    
    def leafNode(node):
        if not node:
            return 
        if not node.left and not node.right:
            kat.append(node.val)
        
        leafNode(node.left)
        leafNode(node.right)
    
    def rightNode(node):
        if not node or not node.left and not node.right:
            return
        
        if node.right:
            rightNode(node.right)
        else:
            rightNode(node.left)
        
        kat.append(node.val)

    leftNode(root.left)
    leafNode(root)
    rightNode(root.right)
    return kat
